java-frontend/src/main/java/org/sonar/java/ast/parser/JavaGrammar.java

```diff
@@ -143,10 +145,24 @@ public CompilationUnitTreeImpl COMPILATION_UNIT() {
           b.token(JavaLexer.SPACING),
           b.optional(PACKAGE_DECLARATION()),
           b.zeroOrMore(IMPORT_DECLARATION()),
+          b.optional(MODULE_DECLARATION()),
           b.zeroOrMore(TYPE_DECLARATION()),
           b.token(JavaLexer.EOF)));
   }
 
+  public ModuleDeclarationTree MODULE_DECLARATION() {
+    return b.<ModuleDeclarationTree>nonterminal(JavaLexer.MODULE_DECLARATION)
+      .is(
+        f.newModuleDeclaration(
+          b.zeroOrMore(ANNOTATION()),
+          b.optional(b.token(JavaRestrictedKeyword.OPEN)),
+          b.token(JavaRestrictedKeyword.MODULE),
+          EXPRESSION_QUALIFIED_IDENTIFIER(),
+          b.token(JavaPunctuator.LWING),
+          // TODO add module directives
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)